unixfs/src/dir/sharded_lookup.rs

```diff
@@ -0,0 +1,460 @@
+    // make a custom Cow type; important not to expose Cow in any API.
+    needle: Cow<'needle, str>,
+}
+
+impl fmt::Debug for ShardedLookup<'_> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            fmt,
+            "ShardedLookup {{ links: {}, needle: {:?} }}",
+            self.links.len(),
+            self.needle.as_ref(),
+        )
+    }
+}
+
+impl<'needle> ShardedLookup<'needle> {
+    /// Returns the next pending link and an iterator over the rest.
+    pub fn pending_links(&self) -> (&Cid, impl Iterator<Item = &Cid>) {
+        let mut iter = self.links.iter();
+        let first = iter.next().expect("Already validated there are links");
+        (first, iter)
+    }
+
+    /// Continues the walk in the DAG of HAMT buckets searching for the original `needle`.
+    pub fn continue_walk(
+        mut self,
+        next: &[u8],
+        cache: &mut Option<Cache>,
+    ) -> Result<MaybeResolved<'needle>, LookupError> {
+        // just to make sure not to mess this up
+        debug_assert_eq!(Some(self.pending_links().0), self.links.front());
+
+        self.links
+            .pop_front()
+            .expect("Already validated there are links");
+
+        let mut hamt = match FlatUnixFs::try_from(next) {
+            Ok(hamt) if hamt.data.Type == UnixFsType::HAMTShard => hamt,
+            Ok(other) => return Err(LookupError::UnexpectedBucketType(other.data.Type.into())),
+            Err(ParsingFailed::InvalidDagPb(e)) | Err(ParsingFailed::InvalidUnixFs(e, _)) => {
+                *cache = Some(self.links.into());
+                return Err(LookupError::Read(Some(e)));
+            }
+            Err(ParsingFailed::NoData(_)) => {
+                *cache = Some(self.links.into());
+                return Err(LookupError::Read(None));
+            }
+        };
+
+        Self::check_supported(&mut hamt)?;
+
+        let found = Self::partition(
+            hamt.links.into_iter(),
+            self.needle.as_ref(),
+            &mut self.links,
+        )?;
+
+        if let Some(cid) = found {
+            *cache = Some(self.links.into());
+            Ok(MaybeResolved::Found(cid))
+        } else if self.links.is_empty() {
+            *cache = Some(self.links.into());
+            Ok(MaybeResolved::NotFound)
+        } else {
+            Ok(MaybeResolved::NeedToLoadMore(self))
+        }
+    }
+
+    /// Transforms this `ShardedLookup` into an `ShardedLookup<'static>` by taking ownership of the
+    /// needle we are trying to find.
+    pub fn with_owned_needle(self) -> ShardedLookup<'static> {
+        let ShardedLookup { links, needle } = self;
+        let needle = Cow::Owned(needle.into_owned());
+        ShardedLookup { links, needle }
+    }
+
+    /// Finds or starts a lookup of multiple buckets.
+    ///
+    /// Returns the found link, the definitive negative or means to continue traversal.
```

Comment:
```suggestion
    /// Returns the found link, the definitive negative or continues the traversal.
```

Reply:
I'll write this as `the means to continue the traversal` but I'll batch your other suggestions.

Reply:
Wasn't 100% sure about the meaning here - feel free to adjust others if I missed it anywhere else.